tests/test_ops.py

```diff
@@ -21,155 +21,195 @@
-    )
-    assert np.allclose(matrix, true_matrix)
-
-
-@pytest.mark.parametrize("phi", phis)
-@pytest.mark.parametrize("mag", mags)
-def test_squeezing_heisenberg(phi, mag):
-    """ops: Tests the Heisenberg representation of the Squeezing gate."""
-    r = mag
-    matrix = cv.Squeezing._heisenberg_rep([r, phi])
-    true_matrix = np.array(
-        [
-            [1, 0, 0],
-            [0, np.cosh(r) - np.cos(phi) * np.sinh(r), -np.sin(phi) * np.sinh(r)],
-            [0, -np.sin(phi) * np.sinh(r), np.cosh(r) + np.cos(phi) * np.sinh(r)],
-        ]
-    )
-    assert np.allclose(matrix, true_matrix)
-
-
-@pytest.mark.parametrize("phi", phis)
-@pytest.mark.parametrize("mag", mags)
-def test_displacement_heisenberg(phi, mag):
-    """ops: Tests the Heisenberg representation of the Displacement gate."""
-    r = mag
-    hbar = 2
-    matrix = cv.Displacement._heisenberg_rep([r, phi])
-    true_matrix = np.array(
-        [
-            [1, 0, 0],
-            [np.sqrt(2 * hbar) * r * np.cos(phi), 1, 0],
-            [np.sqrt(2 * hbar) * r * np.sin(phi), 0, 1],
-        ]
-    )
-    assert np.allclose(matrix, true_matrix)
-
-
-@pytest.mark.parametrize("phi", phis)
-@pytest.mark.parametrize("theta", phis)
-def test_beamsplitter_heisenberg(phi, theta):
-    """ops: Tests the Heisenberg representation of the Beamsplitter gate."""
-    matrix = cv.Beamsplitter._heisenberg_rep([theta, phi])
-    true_matrix = np.array(
-        [
-            [1, 0, 0, 0, 0],
+class TestQubit:
+    """Tests the qubit based operations."""
+
+    identity = np.array([[1, 0], [0, 1]])
+    paulix = np.array([[0, 1], [1, 0]])
+    pauliy = np.array([[0, -1j], [1j, 0]])
+    pauliz = np.array([[1, 0], [0, -1]])
+    hadamard = 1/np.sqrt(2)*np.array([[1, 1],[1, -1]])
+
+    observables = [identity, paulix, pauliy, pauliz, hadamard]
+
+
+    @pytest.mark.parametrize("observable, eigvals, eigvecs", [
+        (np.array([[1, 0], [0, 1]]), np.array([1., 1.]), np.array([[1., 0.],[0., 1.]])),
+        (np.array([[0, 1], [1, 0]]), np.array([-1, 1]), np.array([[-0.70710678,  0.70710678],[ 0.70710678,  0.70710678]])),
+        (np.array([[0, -1j], [1j, 0]]), np.array([-1, 1]), np.array([[-0.70710678+0.j        , -0.70710678+0.j        ], [ 0.        +0.70710678j,  0.        -0.70710678j]])),
+        (np.array([[1, 0], [0, -1]]), np.array([-1, 1]), np.array([[0., 1.], [1., 0.]])),
+        (1/np.sqrt(2)*np.array([[1, 1],[1, -1]]), np.array([-1, 1]), np.array([[ 0.38268343, -0.92387953],[-0.92387953, -0.38268343]])),
+    ])
+    def test_hermitian_eigvals_eigvecs(self, observable, eigvals, eigvecs, tol):
+        """Tests that the eigvals method of the Hermitian class returns the correct results."""
+        key = tuple(observable.flatten().tolist())
+
+        qubit.Hermitian.eigvals(observable)
+        assert np.allclose(qubit.Hermitian._eigs[key]["eigval"], eigvals, atol=tol, rtol=0)
+        assert np.allclose(qubit.Hermitian._eigs[key]["eigvec"], eigvecs, atol=tol, rtol=0)
```

Comment:
awesome!

Reply:
I wonder if we should test creating two Hermitian observables, and checking that `len(Hermitian._eig) == 2`.

Reply:
I've also thought about this, one thing is that ``qubit.Hermitian._eigs`` will not be defaulted to an empty dictionary by default between test runs.

In these tests, this can be solved by ``qubit.Hermitian._eigs = {}`` which I'll include and which makes sense, as we wouldn't like to leave state after each unit test. This will also result in potential failures of tests, should they store the eigendecomposition at different parts have a larger dictionary due to the entries that have previously been stored.

Reply:
Hmm, do you think this could cause memory issues in general use?

For the tests, this sounds like something that would be perfect as a tear down fixture, for example
```python
@pytest.fixture
def tear_down_hermitian():
    yield None
    qubit.Hermitian._eigs = {}

@pytest.mark.usefixtures("tear_down_hermitian")
class TestEigs:
```

Reply:
It could be beneficial as indeed even if the eigendecomposition of the same matrix would be needed, then it would not be computed more than once.

Hmm, I wouldn't see a reason when it could cause memory issues, do you think it could have downsides?

Cool, thanks for the fixture!